src/compute/src/render/continual_task.rs

```diff
@@ -632,91 +629,166 @@ impl<D: Ord> SinkState<D, Timestamp> {
+    /// ```
+    fn times_extract(&self, name: &str) -> (Collection<G, D, R>, Collection<G, (), R>);
+}
+
+impl<G, D, R> TimesExtract<G, D, R> for Collection<G, D, R>
 where
     G: Scope<Timestamp = Timestamp>,
+    D: Clone + 'static,
     R: Semigroup + 'static + std::fmt::Debug,
 {
-    let name = format!("ct_times_reduce({})", name);
-    input
-        .inner
-        .unary_frontier(Pipeline, &name, |_caps, _info| {
-            let mut notificator = FrontierNotificator::new();
-            let mut stash = HashMap::<_, R>::new();
-            let mut buf = Vec::new();
-            move |input, output| {
+    fn times_extract(&self, name: &str) -> (Collection<G, D, R>, Collection<G, (), R>) {
+        let name = format!("ct_times_extract({})", name);
+        let mut builder = OperatorBuilder::new(name, self.scope());
+        let (mut passthrough, passthrough_stream) = builder.new_output();
+        let (mut times, times_stream) = builder.new_output();
+        let mut input = builder.new_input(&self.inner, Pipeline);
+        builder.set_notify(false);
+        builder.build(|_caps| {
+            // We need to drain but iteration order doesn't matter. A BTreeMap
+            // would also be fine but slightly wasteful.
+            #[allow(clippy::disallowed_types)]
+            let mut times_hash = std::collections::HashMap::<_, R>::new();
+
+            let mut passthrough_buf = Vec::new();
+            let mut times_buf = Vec::new();
+            move |_frontiers| {
+                let mut passthrough = passthrough.activate();
+                let mut times = times.activate();
                 while let Some((cap, data)) = input.next() {
-                    data.swap(&mut buf);
-                    for ((), ts, diff) in buf.drain(..) {
-                        notificator.notify_at(cap.delayed(&ts));
-                        if let Some(sum) = stash.get_mut(&ts) {
-                            sum.plus_equals(&diff);
+                    data.swap(&mut passthrough_buf);
+                    for (_data, ts, diff) in &passthrough_buf {
+                        if let Some(d) = times_hash.get_mut(ts) {
+                            d.plus_equals(diff);
                         } else {
-                            stash.insert(ts, diff);
-                        }
+                            times_hash.insert(*ts, diff.clone());
+                        };
                     }
+                    let times_iter = times_hash
+                        .drain()
+                        // Silly to emit zero diffs.
+                        .filter(|(_, diff)| !diff.is_zero())
+                        .map(|(ts, diff)| ((), ts, diff));
+                    times_buf.extend(times_iter);
+                    passthrough
+                        .session(&cap)
+                        .give_container(&mut passthrough_buf);
+                    times.session(&cap).give_container(&mut times_buf);
                 }
```

Comment:
I think this could use the `ConsolidatingContainerBuilder` instead of maintaining its own hashmap, but it might change semantics. For an example, see [flat_map](https://github.com/MaterializeInc/materialize/blob/de0d5dd58b8fdd0138b67c05f5182d60418e3849/src/compute/src/render/flat_map.rs#L126). The consolidating container builder sits on a buffer of fixed length, appends, and whenever full or time ticks, it'll consolidate the buffer. This only ensures opportunistically consolidated outputs, so duplicate elements might still occur.

In this case, the stream seems to be `((), time, diff)`, with `time` being totally ordered. It seems this would be the best case example for the consolidating builder, unless you need potentially zero or more timestamps per input batch. (Note that the implementation emits to `times` on every `input.next()` iteration.)

Reply:
Yup, just `give` and it'll compact in-place.